src/subword.rs

```diff
@@ -81,6 +81,7 @@ pub trait SubwordIndices {
     ///
     /// The largest possible bucket exponent is 64.
     fn subword_indices(&self, min_n: usize, max_n: usize, buckets_exp: usize) -> Vec<u64>;
+    fn ngrams_indices(&self, min_n: usize, max_n: usize, buckets_exp: usize) -> Vec<(String, u64)>;
```

Comment:
Add a new line and rustdoc to document this method.

Since we have a reference to the string itself (`&self`), it would be nicer and more efficient (fewer allocations) to return string references: `Vec<(&str, u64)>`

Reply:
I don't think `&str` is possible as return type due to the `Vec<char>` being owned by the method. 

I would prefer to get `Vec<(Vec<char>, u64)>` as a return type from this method, since we can get mismatches between the length of `String` vs. `Vec<char>` otherwise and we extract ngrams based on the length of `&[char]`.

Reply:
> I don't think `&str` is possible as return type due to the `Vec<char>` being owned by the method.

`NGrams` was made to work with any data type, but that is not really necessary in this crate. We could specialize n-grams to take `&str` and return an iterator over `&str`. We only have to ensure that hashing is still carried out the same way, but it is possible to iterator over a string character-wise.

Reply:
But iirc, slicing `&str`s results in panics if invalid UTF-8 is returned? So just incrementing the beginning of the slice 1-by-1 might lead to crashes, unless I'm missing something.

Reply:
You wouldn't have to increment the beginning one by one. Use `char_indices` to get the offsets of each character in the string and you can make slices that are always valid UTF-8.

Reply:
Though, with all these changes, it would be good to have some benchmarks as well. So, I propose three PRs:

1. Add a benchmark for `NGrams`.
2. Change the `NGrams` implementation to return `&str`. Use (1) to ensure that we are not (much) slower.
3. This PR (rebased on (2)).

Reply:
Thanks for the explanation, I wasn't aware of `char_indices`. Is there also a cheap way to get the number of `char`s in a `&str` without collecting `chars()`?

In that case, we'd have to implement `Iterator` for `NGrams<'a, &str>` and `NGram<'a, &[char]>` since specialization isn't stable yet?

Reply:
> Is there also a cheap way to get the number of `char`s in a `&str` without collecting `chars()` (or `char_indices`)?

Nope, you have to decode in linear time due to variable-length encoding.

Reply:
@NianhengWu In `master` `NGrams` returns string slices, so all should be set up to modify this PR to return `Vec<(&str, u64)>`.

Reply:
yes i noticed in `master`, `NGrams`  just changed, should I close this PR and start a new one?

Reply:
> yes i noticed in `master`, `NGrams` just changed, should I close this PR and start a new one?

You could rebase this branch on master and incorporate the changes or do a new PR. Whichever you find more convenient ;).